init dev

rainset opened the pull request.

rainset committed:
init dev

rainset committed:
fix linter check errors

rainset committed:
поменял структуру настроек
добавил BuildVersion

kilchik commented on cmd/client/main.go:
```diff
@@ -0,0 +1,52 @@
+	defer cancel()
+
+	if _, err := os.Stat(cfg.ClientFolder); errors.Is(err, os.ErrNotExist) {
+		err := os.MkdirAll(cfg.ClientFolder, os.ModePerm)
```
0o777 - явно неподходящее разрешение для директории с секретами - любой пользователь любой группы может их прочитать и перетереть

kilchik commented on cmd/client/main.go:
```diff
@@ -0,0 +1,52 @@
+		}
+	}
+
+	a := app.New(ctx, cfg)
```
Не увидел использования ctx в main. Ctx должен передавать в Run, при получении сигнала из ctx.Done() Run должен корректно завершаться. Сейчас сигнал ОС просто будет проигнорирован, приложение завершится, не дождавшись завершения всех производимых работ

kilchik commented on internal/server/handler/handler.go:
```diff
@@ -0,0 +1,652 @@
+
+	err = c.BindJSON(&rb)
+	if err != nil {
+		logger.Error(err)
```
Все логи должны содержать все необходимые поля с деталями (user_id и пр, но не чувствительные данные), а также должны быть снабжены префиксом, позволяющим однозначно локализовать ошибку по логам

lostsequence replied:
этот комментарий не поправлен

rainset replied:
сделал более информативные логи

kilchik commented on internal/server/handler/handler.go:
```diff
@@ -0,0 +1,652 @@
+
+	rb.UserID = userID
+
+	err = h.service.SaveCard(rb)
```
Выходит все данные хранятся в базе без шифрования, и в случае слива БД злоумышленник сможет воспользоваться всеми сохраненными банковскими картами? Так быть не должно, в БД любая чувствительная информация должна хоть как-то обфусцирована/зашифрована

kilchik commented on internal/server/storage/storage.go:
```diff
@@ -0,0 +1,279 @@
+func (d *Database) SaveCard(card model.DataCard) (err error) {
+	if card.ID == 0 {
+		sql := "INSERT INTO data_cards (user_id,title,number,date,cvv,meta,updated_at) VALUES ($1,$2,$3,$4,$5,$6,$7)"
+		_, err = d.pgx.Exec(d.ctx, sql, card.UserID, card.Title, card.Number, card.Date, card.Cvv, card.Meta, time.Now())
```
Контекст во всех запросах к базе должен браться из контекста хэндлера. Нет никакого смысла передавать context.Background в такие запросы. Контекст в запросе к БД нужен для того, чтобы при обрыве соединения/таймауте подвисший запрос мог завершиться, а вместе с ним и горутина с хэндлером.
Представь, что БД сильно нагружена, не справляется с запросами. Вопреки тому, что клиенты будут отваливаться от сервера по таймауту, соединения с БД будут продолжать висеть, пока не исчерпается их лимит. В итоге несколько долгих запросов подвесят все приложение

kilchik commented on internal/client/app/app.go:
```diff
@@ -0,0 +1,1445 @@
+			cardData.LocalID = item.LocalID
+		}
+
+		err = a.AddCard(cardData)
```
Локально данные тоже не должны храниться в plain text

kilchik commented on cmd/server/main.go:
```diff
@@ -0,0 +1,30 @@
+package main
```
А где тесты? По заданию 80% кода должно быть покрыто тестами

lostsequence replied:
попробуй пожалуйста добавить еще тесты для хендлеров

rainset replied:
добавил

kilchik commented on internal/client/service/http.go:
```diff
@@ -0,0 +1,224 @@
+	}
+}
+
+func (s *HTTPService) DeleteCard(accessToken string, extID int) (err error) {
```
А где методы для сохранения карты и пр данных? Как данные передаются на сервер клиентом?

kilchik commented on internal/client/service/http.go:
```diff
@@ -0,0 +1,224 @@
+func (s *HTTPService) SignIn(user model.User) (tokens model.Tokens, err error) {
+	client := resty.New()
+
+	res, err := client.R().
```
Все данные должны передаваться по https - это же все-таки пароли/банк карты

kilchik commented on README.md:
```diff
@@ -0,0 +1,62 @@
+# Менеджер паролей GophKeeper
```
Я бы ожидал в README описания того, что было выполнено из этих требований и немножко про выбранные технологии и подходы

rainset committed:
- https
- симметричное шифрование данных ключем пользователя
- переработал методы сохранения данных
- добавил синхронизацию

rainset committed:
- правки синхронизации с сервером

rainset committed:
- unit тесты

rainset committed:
- unit тесты

lostsequence commented on internal/server/app/app.go:
```diff
@@ -0,0 +1,94 @@
+
+// Run Инициализация приложения.
+func Run(cfg *config.Config) {
+
```
пустая строка

rainset replied:
убрал

lostsequence commented on internal/server/app/app.go:
```diff
@@ -0,0 +1,94 @@
+// Run Инициализация приложения.
+func Run(cfg *config.Config) {
+
+	ctx := context.Background()
```
пустой контекст особого смысла передавать нет
можно сделать withCancel и cancel вызвать после <-quit
так же можешь рассмотреть конструкцию NotifyContext

rainset replied:
поменял контекст

lostsequence commented on internal/server/app/app.go:
```diff
@@ -0,0 +1,94 @@
+			}
+
+			time.Sleep(60 * time.Second)
+		}
```
по хорошему надо бы подождать горутину, чтобы удаление завершилось штатно на завершение приложения. Можно реализовать с помощью wait group
можно селектом слушать контекст и на <-done прервать цикл и перевести вейт группу в done

rainset replied:
добавил waitGroup

lostsequence commented on internal/server/app/app.go:
```diff
@@ -0,0 +1,94 @@
+
+	ctx := context.Background()
+
+	store := storage.New(ctx, cfg.DatabaseDsn)
```
на завершение приложения коннекшн к базе тоже нужно закрыть

rainset replied:
добавил

lostsequence commented on internal/client/service/http_test.go:
```diff
@@ -0,0 +1,77 @@
+)
+
+func TestHTTPService_AddCard(t *testing.T) {
+	t.Skipped()
```
почему решил скипнуть ?

rainset replied:
добавил тесты для хендлеров сервера, в них думаю тоже самое, не стал делать

lostsequence commented on internal/server/storage/storage.go:
```diff
@@ -0,0 +1,293 @@
+	DeleteText(ctx context.Context, textID, userID int) error
+	FindText(ctx context.Context, textID, userID int) (text model.DataText, err error)
+	FindAllTexts(ctx context.Context, userID int) (texts []model.DataText, err error)
+}
```
обычно таких больших интерфейсов стараются избегать
в целом интерфейсы хорошо объявлять в месте их использования (тогда сервисы тоже надо будет разбить  по их зонам ответственности), таким образом каждый сервис будет использовать интерфейс с заметно меньшим количеством методов.
Это можно не исправлять, как совет.

rainset replied:
понял, изначально не планировал приложение с несколькими сущностями для данных, времени разделить уже нет)
а так можно было разнести интерфейсы по типу данных

rainset committed:
- доработал логирование
- тесты для хендлеров
- правки по ревью

lostsequence commented on internal/server/app/app.go:
```diff
@@ -0,0 +1,102 @@
+				logger.Error(err)
+				return
+			}
+			time.Sleep(60 * time.Second)
```
вместо time.sleep лучше ticker

lostsequence commented on internal/server/app/app.go:
```diff
@@ -0,0 +1,102 @@
+	go func() {
+		defer wg.Done()
+		for {
+			err := newService.ClearExpiredRefreshTokens(ctx)
```
тут надо еще слушать отмену контекста иначе wg.Done() не отработает

lostsequence commented on internal/client/app/app.go:
```diff
@@ -0,0 +1,1930 @@
+			logger.Error(err)
+		}
+
+		//log.Println("id", id)
```
закоменченный код

lostsequence commented on internal/server/app/app.go:
```diff
@@ -0,0 +1,102 @@
+}
+
+func (s *Server) Run() error {
+	return s.httpServer.ListenAndServeTLS("./cert/cert.pem", "./cert/private.key")
```
лучше пути к сертам дать возможность задавать пользователю
